fdbserver/DiskQueue.actor.cpp

```diff
@@ -464,20 +496,19 @@ class RawDiskQueue_TwoFiles {
 			}
 
 			// Read the first pages
-			memset(firstPage, 0, sizeof(Page)*2);
 			vector<Future<int>> reads;
 			for(int i=0; i<2; i++)
 				if( self->files[i].size > 0)
-					reads.push_back( self->files[i].f->read( &firstPage[i], sizeof(Page), 0 ) );
+					reads.push_back( self->files[i].f->read( self->firstPages[i], sizeof(Page), 0 ) );
 			wait( waitForAll(reads) );
 
 			// Determine which file comes first
-			if ( compare( &firstPage[1], &firstPage[0] ) ) {
-				std::swap( firstPage[0], firstPage[1] );
+			if ( compare( self->firstPages[1], self->firstPages[0] ) ) {
+				std::swap( self->firstPages[0], self->firstPages[1] );
 				std::swap( self->files[0], self->files[1] );
 			}
 
-			if ( !compare( &firstPage[1], &firstPage[1] ) ) {
+			if ( !compare( self->firstPages[1], self->firstPages[1] ) ) {
```

Comment:
Dumb question: why is it compare the same piece of memory?

Reply:
`compare` is

```
  // isValid(firstPage) == compare(firstPage, firstPage)
  // isValid(otherPage) == compare(firstPage, otherPage)
  // Swap file1, file2 if comparePages( file2.firstPage, file1.firstPage )
  static bool comparePages( void* v1, void* v2 ) {
    Page* p1 = (Page*)v1; Page* p2 = (Page*)v2;
    return p2->checkHash() && (p2->seq >= p1->seq || !p1->checkHash());
  }
```

So this isn't a straight compare, it's a validity check AND compare.  So `comparePages(x, x)` will fail if `x` is corrupt or invalid.  This code is arranged this way because `RawDiskQueue_TwoFiles` doesn't know anything about the contents or validity of pages.  The only thing it has is a `compare` function that is passed in, so there was a seemingly additional constraint added to compare that `x != x` iff `x` is invalid or corrupt.

Reply:
Hmm, in this case, instead of `compare(firstPage, firstPage)`, why not just call `firstPage.checkHash()` directly?

Reply:
firstPage[0] and firstPage[1] is just a 4K chunk of memory to `RawDiskQueue_TwoFiles`.  A `DiskQueue` knows that the memory is actually a `DiskQueue::Page`, and can call `checkHash()`, but `RawDiskQueue_TwoFiles` doesn't know that.

I'll preemptively agree that the layering here is a bit weird.  I understand that the intention was that a `RawDiskQueue_TwoFiles` presents the "infinite queue of bytes" abstraction, and `DiskQueue` flavors it with pages, hashes, batching pushes, and other extra niceties, but the separation of `RDQ_TF` and `DiskQueue` sometimes results in weird things because a `RDQ_TF` has no ability to understand data that's in a page or its validity.